src/DependencyGraph/DependencyGraphHelpers.js

```diff
@@ -8,26 +8,83 @@
-  constructor({ providesModuleNodeModules, assetExts }) {
-    this._providesModuleNodeModules = providesModuleNodeModules;
+  constructor({ roots, providesModuleNodeModules, assetExts }) {
+    this._roots = roots;
+    this._providesModuleNodeModuleDirs = this._resolveHastePackages(providesModuleNodeModules);
     this._assetExts = assetExts;
   }
 
+  /**
+   * An item in the `providesModuleNodeModule` array is an object containing two keys:
+   *  * "name" - the package name
+   *  * "parent" - if a string, it's the name of the parent package of the module in
+   *  the dep tree. If null, it signifies that we should look for this package at
+   *  the top level of the module tree.
+   *
+   * We try to resolve the specified module...but if we can't, we fallback to looking
+   * in "node_modules/[package name]" (this satisfies use cases where the actual
+   * "react-native" package is one of the root dirs, such as when we run tests
+   * or examples).
+   */
+  _resolveHastePackages(packages) {
+    const _packagePathForPackage = ({ name, parent }, rootDir) => {
+      if (parent) {
+        if (!Array.isArray(parent)) {
+          parent = [parent];
+        }
+        parent.push(name);
+        return path.join(rootDir, 'node_modules', parent.join(`${path.sep}node_modules${path.sep}`)) + path.sep;
+      } else {
+        return path.join(rootDir, 'node_modules', name) + path.sep;
+      }
+    };
+
+    return packages.reduce((hastePackages, p) => {
+      this._roots.forEach(rootDir => {
+        const packagePath = _packagePathForPackage(p, rootDir);
+        try {
+          const stats = fs.statSync(packagePath);
+          if (stats && stats.isDirectory()) {
+            hastePackages.push(packagePath);
+          }
+        } catch (e) {
+          // if we don't find the package, let's just default to node_modules/[package name]
+          hastePackages.push(_packagePathForPackage({ name: p.name }, rootDir));
+        }
+      });
+      return hastePackages;
+    }, []);
+  }
+
+  /**
+   * This method has three possible outcomes:
+   *
+   * 1) A file is not in "node_modules" at all (some type of file in the project).
+   * 2) The file is in "node_modules", and it's contained in one of the
+   * "providesModuleNodeModules" packages.
+   * 3) It's in "node_modules" but not in a "providesModuleNodeModule" package,
+   * so it's just a normal node_module.
+   */
   isNodeModulesDir(file) {
-    const index = file.lastIndexOf('/node_modules/');
+    const index = file.indexOf('/node_modules/');
     if (index === -1) {
       return false;
     }
 
-    const parts = file.substr(index + 14).split(path.sep);
-    const dirs = this._providesModuleNodeModules;
-    for (let i = 0; i < dirs.length; i++) {
-      if (parts.indexOf(dirs[i]) > -1) {
-        return false;
+    const isHaste = this._providesModuleNodeModuleDirs.map(dir => {
+      if (file.startsWith(dir) &&
+          !file.startsWith(path.join(dir, 'node_modules'))
```

Comment:
I'd like to avoid calling `path.join` here for perf reasons. Can we make sure that `dir` never ends in a path-separator in `_resolveHastePackages` and use `+ path.sep +` instead?